File: backend/internal/domain/calculator.go
package domain

const (
	OperationAdd        = "add"
	OperationSubtract   = "subtract"
	OperationMultiply   = "multiply"
	OperationDivide     = "divide"
	OperationPower      = "power"
	OperationSquareRoot = "sqrt"
	OperationPercentage = "percentage"
)

type Calculation struct {
	Operand1  *float64
	Operand2  *float64
	Operation string
}

type CalculatorService interface {
	Add(a, b float64) float64
	Subtract(a, b float64) float64
	Multiply(a, b float64) float64
	Divide(a, b float64) (float64, error)
	Power(base, exponent float64) float64
	SquareRoot(a float64) (float64, error)
	Percentage(value, percent float64) float64
	Execute(Calculation) (float64, error)
}

File: frontend/src/types/calculator.ts
// Request/Response types for API communication
export interface CalculateRequest {
    operand1?: number;
    operand2?: number;
    operation: CalculatorOperation;
}

export interface CalculateResponse {
    result: number;
    error?: string;
}

export type CalculatorOperation =
    | 'add'
    | 'subtract'
    | 'multiply'
    | 'divide'
    | 'power'
    | 'sqrt'
    | 'percentage';

export interface CalculatorState {
    display: string;
    previousValue: number | null;
    operation: CalculatorOperation | null;
    newNumber: boolean;
    isLoading: boolean;
    error: string | null;
}
